src/coreComponents/common/Span.hpp

```diff
@@ -0,0 +1,244 @@
+/*
+ * ------------------------------------------------------------------------------------------------------------
+ * SPDX-License-Identifier: LGPL-2.1-only
+ *
+ * Copyright (c) 2018-2020 Lawrence Livermore National Security LLC
+ * Copyright (c) 2018-2020 The Board of Trustees of the Leland Stanford Junior University
+ * Copyright (c) 2018-2020 Total, S.A
+ * Copyright (c) 2019-     GEOSX Contributors
+ * All rights reserved
+ *
+ * See top level LICENSE, COPYRIGHT, CONTRIBUTORS, NOTICE, and ACKNOWLEDGEMENTS files for details.
+ * ------------------------------------------------------------------------------------------------------------
+ */
+
+/**
+ * @file Span.hpp
+ */
+
+#ifndef GEOSX_COMMON_SPAN_HPP
+#define GEOSX_COMMON_SPAN_HPP
+
+#include "codingUtilities/traits.hpp"
+#include "common/Logger.hpp"
+
+#include <memory>
+#include <iterator>
+#include <type_traits>
+
+namespace geosx
+{
+
+/**
+ * @brief Lightweight non-owning wrapper over a contiguous range of elements.
+ * @tparam T type of range element
+ *
+ * This is a simplified version of C++17's std::span<> that doesn't support
+ * compile-time extents.
+ */
+template< typename T >
+class Span
+{
+public:
+
+  /// Type of range element
+  using element_type = T;
+
+  /// Type of underlying value
+  using value_type = std::remove_cv_t< T >;
+
+  /// Type used for indexing the range
+  using size_type = std::ptrdiff_t;
```

Comment:
Actually `size_type` should preferably be a `std::size_t`, shouldn't it?

Reply:
I made it signed intentionally, because our Array types use `localIndex` (which is a signed type) as size type, and this just makes things like `labels.size() == array.size( dim )` easier to write.

In general, there seems to be an increasing understanding among c++ committee people that unsigned size and indexing were probably a mistake. It's the reason why [std::ssize](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p1227r1.html) was introduced. There is also some rationale given here: https://github.com/ericniebler/stl2/issues/182#issuecomment-257000171

If we want to strictly stick to `std::span`, I should make this `std::size_t`... and add static casts everywhere I use it. Do we want this, is the question?

Reply:
I would preferably stick to the standard so we can replace `geosx::Span` by the `std` version with less risks.

And since the `std` solution seems to use a `std::ssize`, do you think adding a simple implementation in GEOSX (like provided in https://en.cppreference.com/w/cpp/iterator/size) could provide the most standard solution?
When we move to C++20, this can be replaced by `std` (we can add a `// LEGACY_C++20` comment inside the function to find them!).

Instead of having casts here and there in the code, would it make sense to replace them by `ssize`?

Reply:
Fair enough, I'll change to `std::size_t`